packages/financial-templates-lib/src/price-feed/URTHPriceFeed.ts

```diff
@@ -0,0 +1,192 @@
+   * @param {String} apiKey apiKey for polygon api
+   * @param {Integer} lookback How far in the past the historical prices will be available using getHistoricalPrice.
+   * @param {Object} networker Used to send the API requests.
+   * @param {Function} getTime Returns the current time.
+   * @param {Number} priceFeedDecimals Number of priceFeedDecimals to use to convert price to wei.
+   * @param {Integer} minTimeBetweenUpdates Min number of seconds between updates. If update() is called again before
+   *      this number of seconds has passed, it will be a no-op.
+   */
+  constructor(
+    private readonly logger: Logger,
+    private readonly index: string,
+    private readonly apiKey: string,
+    private readonly lookback: number,
+    private readonly networker: NetworkerInterface,
+    private readonly getTime: () => Promise<number>,
+    private readonly priceFeedDecimals = 18,
+    private readonly minTimeBetweenUpdates = 43200 // 12 hours is a reasonable default since this pricefeed returns daily granularity at best.
+  ) {
+    super();
+
+    this.uuid = `Polygon-${index}`;
+
+    this.priceHistory = [];
+
+    this.convertPriceFeedDecimals = (number) => {
+      // Converts price result to wei
+      // returns price conversion to correct decimals as a big number.
+      // Note: Must ensure that `number` has no more decimal places than `priceFeedDecimals`.
+      return Web3.utils.toBN(
+        parseFixed(number.toString().substring(0, priceFeedDecimals), priceFeedDecimals).toString()
+      );
+    };
+  }
+  // Updates the internal state of the price feed. Should pull in any async data so the get*Price methods can be called.
+  // Will use the optional ancillary data parameter to customize what kind of data get*Price returns.
+  // Note: derived classes *must* override this method.
+  // Note: Eventually `update` will be removed in favor of folding its logic into `getCurrentPrice`.
+  public async update(ancillaryData?: string): Promise<void> {
+    const currentTime = await this.getTime();
+
+    // Return early if the last call was too recent.
+    if (this.lastUpdateTime !== null && this.lastUpdateTime + this.minTimeBetweenUpdates > currentTime) {
+      this.logger.debug({
+        at: "URTHPriceFeed",
+        message: "Update skipped because the last one was too recent",
+        currentTime: currentTime,
+        lastUpdateTimestamp: this.lastUpdateTime,
+        timeRemainingUntilUpdate: this.lastUpdateTime + this.minTimeBetweenUpdates - currentTime,
+      });
+      return;
+    }
+
+    this.logger.debug({
+      at: "URTHPriceFeed",
+      message: "Updating URTHPriceFeed",
+      currentTime: currentTime,
+      lastUpdateTimestamp: this.lastUpdateTime,
+    });
+
+    // 1. Construct URL.
+    // See https://polygon.io/docs/stocks/getting-started.
+    const url = "https://api.polygon.io/v1/open-close/" + this.index + `2023-01-09?adjusted=true&apiKey=${this.apiKey}`;
```

Comment:
The way the update function works is a bit complex. It basically runs at regular intervals, decided by the `minTimeBetweenUpdates` variable. The update function then get's a historical response from the API which is between 2 timestamps decided by `startDateString` and `endDateString`. The start date depends on the `lookback` variable.

The current price is stored in the `currentPrice` variable which is retrieved from the historical prices that the API sent back.

In the above code, the date is hardcoded to `2023-01-09` which is incorrect, it should instead use the dates I explained above.

Reply:
You can take inspiration on how to implement above from `CommoditiesApiPriceFeed.ts` file.